File: MossTaps/data/Java_Students/Project1-tca3-java/Agent.java
package ravensproject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

// Uncomment these lines to access image processing.
//import java.awt.Image;
//import java.io.File;
//import javax.imageio.ImageIO;

/**
 * Your Agent for solving Raven's Progressive Matrices. You MUST modify this
 * file.
 * 
 * You may also create and submit new files in addition to modifying this file.
 * 
 * Make sure your file retains methods with the signatures: public Agent()
 * public char Solve(RavensProblem problem)
 * 
 * These methods will be necessary for the project's main method to run.
 * 
 */
public class Agent {

	public static int MAX_ANGLE = 360;
	public static int MAX_REFLECTION_SPLITS = 4; // maximum number of
													// reflections which the
													// agent will attempt to
													// match numbers

	public static HashMap<String, String> BINARY_PAIRS;
	public static String[] BinaryPairs1 = { "top", "left" };
	public static String[] BinaryPairs2 = { "bottom", "right" };

	public CaseBasedReasoningSystem mCaseBasedReasoningSystem;
	public ArrayList<ObjectPairing> mMostRecentObjectPairings;

	/**
	 * The default constructor for your Agent. Make sure to execute any
	 * processing necessary before your Agent starts solving problems here.
	 * 
	 * Do not add any variables to this signature; they will not be used by
	 * main().
	 * 
	 */
	public Agent() {
		BINARY_PAIRS = new HashMap<String, String>();
		for (int i = 0; i < BinaryPairs1.length; i++) {
			BINARY_PAIRS.put(BinaryPairs1[i], BinaryPairs2[i]);
			BINARY_PAIRS.put(BinaryPairs2[i], BinaryPairs1[i]);
		}

		mCaseBasedReasoningSystem = new CaseBasedReasoningSystem();
		mMostRecentObjectPairings = new ArrayList<ObjectPairing>();

	}

	/**
	 * The primary method for solving incoming Raven's Progressive Matrices. For
	 * each problem, your Agent's Solve() method will be called. At the
	 * conclusion of Solve(), your Agent should return a String representing its
	 * answer to the question: "1", "2", "3", "4", "5", or "6". These Strings
	 * are also the Names of the individual RavensFigures, obtained through
	 * RavensFigure.getName().
	 * 
	 * In addition to returning your answer at the end of the method, your Agent
	 * may also call problem.checkAnswer(String givenAnswer). The parameter
	 * passed to checkAnswer should be your Agent's current guess for the
	 * problem; checkAnswer will return the correct answer to the problem. This
	 * allows your Agent to check its answer. Note, however, that after your
	 * agent has called checkAnswer, it will *not* be able to change its answer.
	 * checkAnswer is used to allow your Agent to learn from its incorrect
	 * answers; however, your Agent cannot change the answer to a question it
	 * has already answered.
	 * 
	 * If your Agent calls checkAnswer during execution of Solve, the answer it
	 * returns will be ignored; otherwise, the answer returned at the end of
	 * Solve will be taken as your Agent's answer to this problem.
	 * 
	 * @param problem
	 *            the RavensProblem your agent should solve
	 * @return your Agent's answer to this problem
	 */
	public int Solve(RavensProblem problem) {

		// declaration of guess
		int myGuess = -1;

		// get problem type
		String problemType = problem.getProblemType();

		// 2x2 verbal problems
		if (problem.hasVerbal() && problemType.equals("2x2")) {
			// attempt to solve verbal representation
			myGuess = SolveVerbal(problem);
		}
		// TODO: Implement other problem types

		int rightAnswer = problem.checkAnswer(myGuess);

		// store information to case-based reasoning system
		mCaseBasedReasoningSystem.AddResult(rightAnswer == myGuess);

		// TODO: Use right answer to affect behavior going forward

		return myGuess;
	}

	// Solves 2x2 verbal problems
	public int SolveVerbal(RavensProblem problem) {
		// create multiple dimension array out of RavensFigures
		HashMap<String, RavensFigure> figuresHash = problem.getFigures();
		RavensFigure[][] figures = new RavensFigure[2][2];

		// iterate through hash map and store figures to array
		int arrayIndex = 0;
		for (char i = 'A'; i < 'D'; i++) {
			// check that hash map contains the desired key
			// if not, return -1
			if (!(figuresHash.containsKey(Character.toString(i)))) {
				System.out.println("Improperly formatted problem");
				return -1;
			}

			figures[arrayIndex % 2][arrayIndex / 2] = figuresHash.get(Character
					.toString(i));
			arrayIndex++;
		}

		// create strings to represent transformations
		TransformList horizontalTransforms = CompareFigures(figures[0][0],
				figures[1][0]);
		TransformList verticalTransforms = CompareFigures(figures[0][0],
				figures[0][1]);

		System.out.println("\n" + problem.getName());
		System.out.println("Horizontal Transform:\n"
				+ horizontalTransforms.toString());
		System.out.println("Vertical Transform:\n"
				+ verticalTransforms.toString());

		// create guess for figure D by applying transforms to existing figure
		RavensFigure guess = ApplyTransforms(figures[0][0],
				horizontalTransforms);
		guess = ApplyTransforms(guess, verticalTransforms);

		int myGuess = 0;
		double myGuessDifference = Double.POSITIVE_INFINITY;
		TransformList myGuessTransformList = null;

		// compare guess to possible answer choices
		for (int i = 1; i <= 6; i++) {
			TransformList transformDifference = CompareFigures(guess,
					figuresHash.get(Integer.toString(i)));
			System.out.printf("Answer Choice: %d %d\n", i,
					transformDifference.GetDifferenceScore());
			// DEBUG System.out.println(transformDifference.toString());
			if (transformDifference.GetDifferenceScore() < myGuessDifference) {
				myGuessDifference = transformDifference.GetDifferenceScore();
				myGuess = i;
				myGuessTransformList = transformDifference;
			}
			// in case of a tie, consult case-based reasoning system to break
			// tie
			else if (mCaseBasedReasoningSystem
					.GetNumSuccesses(transformDifference) > mCaseBasedReasoningSystem
					.GetNumSuccesses(myGuessTransformList)) {
				myGuessDifference = transformDifference.GetDifferenceScore();
				myGuess = i;
				myGuessTransformList = transformDifference;
			}
		}

		return myGuess;
	}

	// returns list of transforms describing difference between inputted figures
	public TransformList CompareFigures(RavensFigure figure1,
			RavensFigure figure2) {
		TransformList result = new TransformList();

		Iterator<RavensObject> it1 = figure1.getObjects().values().iterator();
		Iterator<RavensObject> it2 = figure2.getObjects().values().iterator();

		// create lists of RavensObjects
		ArrayList<RavensObject> list1 = new ArrayList<RavensObject>();
		ArrayList<RavensObject> list2 = new ArrayList<RavensObject>();
		while (it1.hasNext()) {
			list1.add(it1.next());
		}

		while (it2.hasNext()) {
			list2.add(it2.next());
		}

		// map objects across figures
		ObjectPairingList objectPairs = new ObjectPairingList();
		for (RavensObject object1 : list1) {
			for (RavensObject object2 : list2) {
				objectPairs.AddObjectPairing(object1, object2,
						CalculateObjectDifferenceScore(object1, object2));
			}
		}

		mMostRecentObjectPairings = objectPairs.GetBestPairings();

		// get difference strings between each object mapping
		for (ObjectPairing objectPairing : mMostRecentObjectPairings) {
			result.AddTransforms(objectPairing.toTransformList());
		}

		return result;
	}

	// returns difference score between two inputted objects
	public int CalculateObjectDifferenceScore(RavensObject object1,
			RavensObject object2) {
		int result = 0;

		HashMap<String, String> hashmap1 = object1.getAttributes();
		HashMap<String, String> hashmap2 = object2.getAttributes();

		for (Map.Entry<String, String> entry : hashmap1.entrySet()) {
			String key = entry.getKey();
			String value = entry.getValue();

			// check if hashmap contains key, if not, add 1 to diff score
			if (hashmap2.containsKey(key)) {
				// check if values are different, if so, add 1 to diff score
				if (!(value.equals(hashmap2.get(key)))) {
					result++;
				}
			} else {
				result++;
			}
		}

		// also check whether object2 has any attributes which object1 does not
		// have
		for (Map.Entry<String, String> entry : hashmap2.entrySet()) {
			String key = entry.getKey();

			if (!(hashmap1.containsKey(key))) {
				result++;
			}
		}

		return result;
	}

	public RavensFigure ApplyTransforms(RavensFigure figure,
			TransformList transforms) {

		// get relevant data out of RavensFigure
		HashMap<String, RavensObject> hashmap = figure.getObjects();

		for (Transform transform : transforms.mList) {

			// get relevant object out of hashmap
			RavensObject object = hashmap.get(transform.mOriginalObjectName);

			// if creation
			if (transform.isCreation()) {
				hashmap.put(transform.mNewObjectName,
						transform.mCreationDeletionObject);
			}

			// if deletion
			else if (transform.isDeletion()) {
				hashmap.remove(transform.mOriginalObjectName);
			}

			// if numeric, apply numeric transform
			else if (transform.isNumeric()) {

				// attempt to apply reflection logic instead of
				// addition/subtraction
				if (transform.GetReflectionLine() != null) {
					String attribute = object.getAttributes().get(
							transform.mAttribute);
					int attributeNum = Integer.parseInt(attribute);
					double transformDifference = 2 * ((transform
							.GetReflectionLine() - attributeNum));
					if (Math.abs(transformDifference) > MAX_ANGLE / 2) {
						transformDifference += MAX_ANGLE;
					}

					attributeNum += transformDifference;
					attributeNum = NormalizeAttributeValue(
							transform.mAttribute, attributeNum); // get number
																	// into
																	// range

					object.getAttributes().put(transform.mAttribute,
							Integer.toString(attributeNum));

				} else {
					String attribute = object.getAttributes().get(
							transform.mAttribute);
					int transformDifference = transform.GetNumericDifference();
					int attributeNum = Integer.parseInt(attribute);

					attributeNum += transformDifference;
					attributeNum = NormalizeAttributeValue(
							transform.mAttribute, attributeNum); // get number
																	// into
																	// range

					object.getAttributes().put(transform.mAttribute,
							Integer.toString(attributeNum));
				}

			}

			else if (object == null) {
				// do nothing
			}
			// otherwise, apply text transform
			else {

				// attempt to apply reflection logic before resorting to direct
				// text
				// replacement
				String attribute;
				if ((transform.getBinaryTextTransform()) != null) {
					String[] textTransform = transform.getBinaryTextTransform();

					attribute = object.getAttributes()
							.get(transform.mAttribute);
					attribute = attribute.replace(textTransform[0],
							textTransform[1]);
					object.getAttributes().put(transform.mAttribute, attribute);
				} else {
					attribute = transform.mNewValue;

					object.getAttributes().put(transform.mAttribute, attribute);
				}
			}

			// TODO: semantic network representation, apply to transform using
			// object mappings
		}
		return figure;
	}

	public int NormalizeAttributeValue(String attributeName, int attributeNum) {
		// if attribute is an angle
		if (attributeName.equals("angle")) {
			return attributeNum % MAX_ANGLE;
		}

		// otherwise, just return the number unchanged
		return attributeNum;
	}

	class CaseBasedReasoningSystem {

		public HashMap<String, Integer> mSuccesses;
		public TransformList mMostRecentCase;

		public CaseBasedReasoningSystem() {
			mSuccesses = new HashMap<String, Integer>();
			mMostRecentCase = new TransformList();
		}

		public void AddCase(TransformList transforms) {
			mMostRecentCase = transforms;
		}

		public void AddResult(boolean result) {
			for (Transform t : mMostRecentCase.mList) {
				if (!(mSuccesses.containsKey(t.mAttribute))) {
					// add case
					mSuccesses.put(t.mAttribute, 0);
				}

				// if successful, increment hash table entry by 1
				if (result) {
					mSuccesses.put(t.mAttribute,
							mSuccesses.get(t.mAttribute) + 1);
				}

			}
		}

		// returns average number of successes with a particular set of
		// transforms
		public double GetNumSuccesses(TransformList transforms) {
			int result = 0;

			for (Transform t : transforms.mList) {
				if (mSuccesses.containsKey(t.mAttribute)) {
					result += mSuccesses.get(t.mAttribute);
				}
			}

			return 1.0 * result / transforms.mList.size();
		}
	}
}

File: MossTaps/data/Current/P1-2015/Berys-Gonzalez, Carolina(a876148095ff319cdf1d6d971f2c0ca5)/Submission attachment(s)/Agent.py
# Your Agent for solving Raven's Progressive Matrices. You MUST modify this file.
#
# You may also create and submit new files in addition to modifying this file.
#
# Make sure your file retains methods with the signatures:
# def __init__(self)
# def Solve(self,problem)
#
# These methods will be necessary for the project's main method to run.

# Install Pillow and uncomment this line to access image processing.
#from PIL import Image

class Agent:
    # The default constructor for your Agent. Make sure to execute any
    # processing necessary before your Agent starts solving problems here.
    #
    # Do not add any variables to this signature; they will not be used by
    # main().
    def __init__(self):
        pass

    # The primary method for solving incoming Raven's Progressive Matrices.
    # For each problem, your Agent's Solve() method will be called. At the
    # conclusion of Solve(), your Agent should return an integer representing its
    # answer to the question: "1", "2", "3", "4", "5", or "6". These integers
    # are also the Names of the individual RavensFigures, obtained through
    # RavensFigure.getName() (as Strings).
    #
    # In addition to returning your answer at the end of the method, your Agent
    # may also call problem.checkAnswer(int givenAnswer). The parameter
    # passed to checkAnswer should be your Agent's current guess for the
    # problem; checkAnswer will return the correct answer to the problem. This
    # allows your Agent to check its answer. Note, however, that after your
    # agent has called checkAnswer, it will *not* be able to change its answer.
    # checkAnswer is used to allow your Agent to learn from its incorrect
    # answers; however, your Agent cannot change the answer to a question it
    # has already answered.
    #
    # If your Agent calls checkAnswer during execution of Solve, the answer it
    # returns will be ignored; otherwise, the answer returned at the end of
    # Solve will be taken as your Agent's answer to this problem.
    #
    # Make sure to return your answer *as an integer* at the end of Solve().
    # Returning your answer as a string may cause your program to crash.
    def Solve(self,problem):
        if not problem.hasVerbal:
            return -1
        def list_of_names_in_figure(fig_objs):
            nlist = []
            for obj in fig_objs:
                nlist.append(obj)
            return nlist
 
        def combo_maker(list1, list2):
            list1 = list_of_names_in_figure(list1)
            list2 = list_of_names_in_figure(list2)
            combinations = []
            if not list1:
                for object2 in list2:
                    combinations.append([None, object2])
                return combinations
            elif not list2:
                for object1 in list1:
                    combinations.append([object1, None])
                return combinations
            else:
                # initialize with all a transformations
                if list1[0]:
                    object_a = list1[0]
                for object2 in list2:
                    combinations.append([(object_a,object2)])
                combinations.append([(object_a, None)])
                change = True
                while change:
                    change = False
                    for obj1 in list1[1:]:
                        for obj2 in list2:
                            for path in combinations:
                                if (obj1, obj2) not in path:
                                    if obj1 not in (i[0] for i in path) and obj2 not in (i[1] for i in path):
                                        path.append((obj1, obj2))
                                        change = True
                                    elif obj2 not in (i[1] for i in path):
                                        path2 = list(path)
                                        for pair in path2:
                                            if pair[0] == obj1:
                                                path2.remove(pair)
                                                path2.append((obj1, obj2))
                                        if path2 not in combinations:
                                            combinations.append(path2)
                                            change = True
                                if (obj1, None) not in path:
                                    if obj1 not in (i[0] for i in path):
                                        path3 = list(path)
                                        path3.append((obj1, None))
                                        if path3 not in combinations:
                                            combinations.append(path3)
                                            change = True
            for fin_path in combinations:
                for end_obj in list2:
                    if end_obj not in (i[1] for i in fin_path):
                        fin_path.append((None, end_obj))
            return combinations

        transform_frame = {"shape": "missing", "size": "missing", "fill": "missing", "angle": "missing", "inside": "missing", "above" : "missing"}

        def transform(obj1, obj2, tframe = transform_frame):
            filled_tframe = dict(tframe)
            def fill_frames(obj):
                objframe = dict(tframe)
                for tkey in obj.attributes.keys():
                    if tkey == "inside" or tkey == "above":
                        attr_val = str(len(obj.attributes[tkey]))
                        objframe[tkey] = attr_val
                    else:
                        objframe[tkey] = obj.attributes[tkey]
                return objframe
            obj1frame = fill_frames(obj1)
            obj2frame = fill_frames(obj2)
            for tkey in tframe.keys():
               if obj1frame[tkey] == obj2frame[tkey]: 
                    filled_tframe[tkey] = "No_change"
               elif tkey == "angle":
                    if obj1frame[tkey] != "missing" and obj2frame[tkey] != "missing":
                        tval = str(abs(int(obj1frame[tkey]) - int(obj2frame[tkey])))
                        filled_tframe[tkey] = tval
                    else:
                        filled_tframe[tkey] = obj1frame[tkey] + "_" + obj2frame[tkey]
               elif obj1frame[tkey] != obj2frame[tkey]:
                    filled_tframe[tkey] = obj1frame[tkey] + "_" + obj2frame[tkey]
            return filled_tframe
            
        def transform_removed(obj1, tframe = transform_frame):
            filled_tframe = dict(tframe)
            for tkey in obj1.attributes.keys():
                filled_tframe[tkey] = obj1.attributes[tkey] + "_missing"
            return filled_tframe

        def transform_new(obj1, tframe = transform_frame):
            filled_tframe = dict(tframe)
            for tkey in obj1.attributes.keys():
                filled_tframe[tkey] = "missing_" + obj1.attributes[tkey] 
            return filled_tframe

        def transform_path(combos, fig1, fig2):
            filled_tpath = []
            for path in combos:
                newtpath = []
                for pair in path:
                    if not pair[1]:
			tr = transform_removed(fig1.objects[pair[0]])
                    elif not pair[0]:
			tr = transform_new(fig2.objects[pair[1]])
                    else:
			tr = transform(fig1.objects[pair[0]], fig2.objects[pair[1]])
                    newtpath.append(tr)
                filled_tpath.append(newtpath)
            return filled_tpath

        A = problem.figures["A"]
        B = problem.figures["B"]
        C = problem.figures["C"]
        comboAB = combo_maker(A.objects, B.objects)
        comboAC = combo_maker(A.objects, C.objects)
        tpathAB = transform_path(comboAB, A, B)
        tpathAC = transform_path(comboAC, A, C)

        def tie_breaker(ans1, ans2, score):
            # [TODO] define tie breaker
            return ans1

        matching_pathBD = []
        matching_pathCD = []
        def check_answer(ans):
            matching_pathBD.append([])
            matching_pathCD.append([])
            score = -1
            D = problem.figures[ans]
            comboBD = combo_maker(B.objects, D.objects)
            comboCD = combo_maker(C.objects, D.objects)
            tpathBD = transform_path(comboBD, B, D)
            tpathCD = transform_path(comboCD, C, D)
            for path in tpathBD:
                if path in tpathAC:
                    score += 2
                    matching_pathBD[int(ans)-1].append(path)
            for path in tpathCD:
                if path in tpathAB:
                    score += 2
                    matching_pathCD[int(ans)-1].append(path)
            return score

        top_score = 0
        top_ans_choice = None
        for ans_int in range(1,7):
            ans_choice = str(ans_int)
            score = check_answer(ans_choice) 
            if score > top_score:
                top_score = score
                top_ans_choice = ans_choice
            elif score == top_score:
                #top_ans_choice = tie_breaker(ans_choice, top_ans_choice, score)
                top_ans_choice = ans_choice

        if top_ans_choice:
            final_answer = int(top_ans_choice)
        else:
            final_answer = -1
        return final_answer
